rosidl_cli/rosidl_cli/cli.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import signal
+
+from rosidl_cli.command.generate import GenerateCommand
+from rosidl_cli.common import get_first_line_doc
+
+
+def add_subparsers(parser, cli_name, commands, dest='_command'):
+    """
+    Create argparse subparser for each command.
+
+    The ``cli_name`` is used for the title and description of the
+    ``add_subparsers`` function call.
+
+    For each command a subparser is created.
+    If the command has an ``add_arguments`` method it is being called.
+
+    :param parser: the parent argument parser
+    :type parser: :py:class:`argparse.ArgumentParser`
+    :param str cli_name: name of the command line command to which the
+      subparsers are being added
+    :param commands: each of the commands contributing specific arguments
+    :type commands: :py:class:`List[Command]`
+    :param str dest: name of the attribute under which the selected
+      command will be stored
+    """
+    # add subparser with description of available subparsers
+    description = ''
+
+    commands = sorted(commands, key=lambda command: command.name)
+    max_length = max(len(command.name) for command in commands)
+    for command in commands:
+        description += '%s  %s\n' % (
+            command.name.ljust(max_length),
+            get_first_line_doc(command))
+    subparser = parser.add_subparsers(
+        title='Commands', description=description,
+        metavar=f'Call `{cli_name} <command> -h` for more detailed usage.')
+    # use a name which doesn't collide with any argument
+    # but is readable when shown as part of the the usage information
+    subparser.dest = ' ' + dest.lstrip('_')
+    subparser.required = True
+
+    # add extension specific sub-parser with its arguments
+    for command in commands:
+        command_parser = subparser.add_parser(
+            command.name,
+            description=get_first_line_doc(command),
+            formatter_class=argparse.RawDescriptionHelpFormatter)
+        command_parser.set_defaults(**{dest: command})
+        if hasattr(command, 'add_arguments'):
```

Comment:
> Any way to avoid the copy of this function?

It is not an exact copy though.

> IIUC rosidl will define all its own commands.

For now :) I subjectively would like not to close doors if it's not strictly necessary. WDYT?

Reply:
> For now :) I subjectively would like not to close doors if it's not strictly necessary. WDYT?

I strongly favor assuming all the commands will live in rosidl now because it will be easier to maintain. The door is always open to make it extensible later if that becomes a requirement.

Not making it extensible now means there will be less code to look through when fixing bugs, it will be easier to read when getting context while reviewing PRs, and it will be easier to refactor when new requirements come. Subjectively, the cost doesn't seem worth it without a good use case. Is there a use case you have in mind?

Reply:
No use case at hand. I'll assume `add_arguments()` is always there.

Reply:
See https://github.com/ros2/rosidl/pull/567/commits/d892fc2a4dbf536b7d1cd383dc7a0f9999d0b9f7.

Reply:
Ah, oops. That changes almost nothing. For some reason I was sure getting rid of the extensibility would get rid a lot more code, but that's definitely not the case.

Reply:
Available commands are explicitly listed already (see [here](https://github.com/ros2/rosidl/pull/567/files#diff-ba5b2eb0619a2f7bae116ef8fcb375c5a7bdd4ef4969394778864fc1136055faR82)),  and I got rid of as much code I could in the original implementation of `add_subparsers`. The latter is just too handy when it comes to adding new commands (#575 introduces `translate`, and `build-configure` is pending).

Reply:
Yeah, I completely misunderstood this file. Thanks for being patient with my noise-making. There are a couple minor things that can be made simpler. 908e948c1b3d31eb2dba9f16f93ab3d6f01f85b2 + e2c1a9c3357c89b7c40aa3d3861c6e9d836262a6 + I'd recommend a small tweak to the comments 8594afde734f1b5750d683597ef9671e2befe624

Reply:
All look good. Cherry-picked them on top of this branch.